### ISSUE TITLE
azcopy sync not handling files/directories with overlapping paths correctly

### ISSUE THREAD

**Issue by pmackles** [Azure/azure-storage-azcopy#148]
Version: AzCopy 10.0.4-Preview
Platform: Linux Ubuntu 16.04.5 LTS (Xenial Xerus)

This is the command:
```
./azcopy sync 'https://somestorageaccount.blob.core.windows.net/logs?<sas>' /mnt/logs --recursive=true --force
```

If you have something like the following in the source container:
```
/logs/foo/bar/log.txt
/logs/foo/bar
```

You'll see failed transfers because the file `/logs/foo/bar` is created on the destination, preventing the creation of the directory `/logs/foo/bar` which is needed to create `/logs/foo/bar/log.txt` on the destination.

You can also see the error in the azcopy log:

```
2019/01/02 13:27:33 ERR: [P#0-T#8] DOWNLOADFAILED: https://somestorageaccount.blob.core.windows.net/logs/foo/bar/log.txt?si=logs-read-only&sig=REDACTED&sr=c&sv=2018-03-28 :
 000 : File Creation Error open /logs/foo/bar/log.txt: not a directory
```

I get that this is a result of going from a flat namespace to a hierarchical namespace. The system generating the blobs on the source account is creating the files. My guess is that its because its using the hadoop `FileSystem` APIs and calling `mkdir()` somewhere along the way so a placeholder is created. While I can try to get them to change the behavior, its another azure service so that is likely another ticket.

Current workaround is to look for these placeholder files (always 0 bytes in length) and delete them before running `azcopy`. Ideally, `azcopy` would also add support for an option to handle this situation by creating the directories first and/or adding an option to ignore 0-length files.

**Comment by zezha-msft** [comment]
Hi @pmackles, thanks for reaching out!

AzCopy should be able to handle this scenario; there might be a bug if it's not working with sync. Could you please help me verify whether `/logs/foo/bar` has the metadata `hdi_isfolder=true`? Thanks!

**Comment by pmackles** [comment]
@zezha-msft - I can confirm that these placeholder blobs (like /logs/foo/bar) have the following metadata key/value pairs:

```
hdi_isfolder=true
hdi_permission={"owner":"root","group":"supergroup","permissions":"rwxrwxrwx"}
```

**Comment by zezha-msft** [comment]
Hi @pmackles, thanks for getting back to us!

It indeed seems that there's a bug in the sync command, we should ignore such blobs. I've logged this item in our backlog. Thank you!

**Comment by zezha-msft** [comment]
@pmackles sorry for the inconvenience, and I really appreciate your patience!

I've rewritten the sync command and released it in 10.0.8. This issue should be fixed. Please let us know if you still encounter any problem.

Here is a [wiki](https://github.com/Azure/azure-storage-azcopy/wiki/Sync-command-guide) explaining how the new sync command works. Please feel free to reach out if you have any question or concern.
